base/uk.ac.stfc.isis.ibex.journal/src/uk/ac/stfc/isis/ibex/journal/JournalSearchParameters.java

```diff
@@ -0,0 +1,179 @@
+
+/*
+ * This file is part of the ISIS IBEX application. Copyright (C) 2012-2019
+ * Science & Technology Facilities Council. All rights reserved.
+ *
+ * This program is distributed in the hope that it will be useful. This program
+ * and the accompanying materials are made available under the terms of the
+ * Eclipse Public License v1.0 which accompanies this distribution. EXCEPT AS
+ * EXPRESSLY SET FORTH IN THE ECLIPSE PUBLIC LICENSE V1.0, THE PROGRAM AND
+ * ACCOMPANYING MATERIALS ARE PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES
+ * OR CONDITIONS OF ANY KIND. See the Eclipse Public License v1.0 for more
+ * details.
+ *
+ * You should have received a copy of the Eclipse Public License v1.0 along with
+ * this program; if not, you can obtain a copy from
+ * https://www.eclipse.org/org/documents/epl-v10.php or
+ * http://opensource.org/licenses/eclipse-1.0.php
+ */
+
+package uk.ac.stfc.isis.ibex.journal;
+
+import java.util.ArrayList;
+import java.util.Calendar;
+import java.util.Optional;
+
+/**
+ * Encapsulates journal search parameters.
+ */
+public class JournalSearchParameters {
+    private JournalField field = JournalField.RUN_NUMBER;
+    private Optional<String> searchString = Optional.empty();
+    private Optional<Integer> fromNumber = Optional.empty();
+    private Optional<Integer> toNumber = Optional.empty();
+    private Optional<Calendar> fromTime = Optional.empty();
+    private Optional<Calendar> toTime = Optional.empty();
+    private ArrayList<JournalSort> sorts = new ArrayList<JournalSort>();
+    {
+        // This is the default sort
+        addSort(JournalField.RUN_NUMBER);
+    }
+    
+    /**
+     * A search can only be made with parameters of one type.
+     * Once some parameters have been set, this variable is set to true
+     * and used to prevent more parameters from being added.
+     */
+    private boolean set = false;
```

Comment:
If this is behaviour you want to stop other developers from doing you should make it impossible for them to do it at compile time. Otherwise this will lead to exceptions during run time. In this case I think that would be best done by splitting this class into four:
* An abstract base class, whose constructor takes a JournalField
* A SearchString class whose constructor takes a String
* A FromTo class whose constructor takes a fromNumber and toNumber 
* A Time class whose constructor takes a fromTime and toTime

You can also then move the logic of what parts to put in the SQL and what to fill the statement with into these classes too. We can discuss further if this doesn't make sense.

Reply:
I've replaced `JournalSearchParameters` and `JournalSqlStatement` with the abstract base class `JournalSearch` and the following subclasses:
- `EmptySearch`
- `SearchString`
- `SearchNumber`
- `SearchTime`

Reply:
This is much improved. Even if it is a bit more verbose it's made it far more extensible.